File: Sample/java/src/main/java/net/nend/sample/java/nativead/NativeMarqueeActivity.java
package net.nend.sample.java.nativead;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.Layout;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import net.nend.android.NendAdNative;
import net.nend.android.NendAdNativeClient;
import net.nend.android.NendAdNativeViewBinder;
import net.nend.sample.java.R;

/**
 * Sample for like telop
 */
public class NativeMarqueeActivity extends AppCompatActivity {

    private String TAG = getClass().getSimpleName();
    private MarqueeView mMarqueeView;
    private TextView mTextView;
    private View mAdContainer;

    private boolean mThreadStopper = false; // for leaving the loop
    private int mScrollViewWidth;           // MarqueeView width
    private int mWidthWithBlank;            // NendNativeTextView width

    private Thread mThread = null;          // runnable thread
    private int mRepeatCount;               // number of repeat
    private int mCurrentX;                  // current text position
    private boolean mIsEndlessLoop;         // endless flag

    /**
     * Runnable variable for marquee
     */
    private Runnable runnable = new Runnable() {
        public void run() {
            // end point
            int endX = getEndX();

            int repeatLimit = 2;
            int textMoveSpeed = 10;

            while (mRepeatCount < repeatLimit) {
                if (!mThreadStopper && !Thread.interrupted()) {

                    mCurrentX = getMarqueeStartX();

                    long beforeTime = System.currentTimeMillis();
                    long afterTime;
                    int fps = 30;
                    long frameTime = 1000 / fps;

                    while (true) {

                        if (!mThreadStopper && !Thread.interrupted()) {

                            if (mCurrentX >= endX) {
                                mRepeatCount++;
                                if (mIsEndlessLoop) {
                                    mRepeatCount--;
                                }
                                break;
                            }

                            mCurrentX += textMoveSpeed;

                            // UI control must on UI thread
                            runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    mMarqueeView.smoothScrollTo(mCurrentX, 0);
                                }
                            });

                            // speed control
                            afterTime = System.currentTimeMillis();
                            long pastTime = afterTime - beforeTime;
                            long sleepTime = frameTime - pastTime;
                            if (sleepTime > 0) {
                                try {
                                    Thread.sleep(sleepTime);
                                } catch (InterruptedException e) {
                                    Log.d(TAG, "thread interruption");
                                }
                            }
                            beforeTime = System.currentTimeMillis();
                        } else if (mThreadStopper) {
                            break;
                        }
                    }
                } else if (mThreadStopper) {
                    break;
                }
            }
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.native_marquee);

        mMarqueeView = (MarqueeView) findViewById(R.id.horizontalScrollView);
        mTextView = (TextView) findViewById(R.id.ad_content);
        mAdContainer = findViewById(R.id.ad);

        final int SPOT_ID = 485516;
        final String API_KEY = "16cb170982088d81712e63087061378c71e8aa5c";

        final NendAdNativeViewBinder mBinder = new NendAdNativeViewBinder.Builder()
                .adImageId(R.id.ad_image)
                .titleId(R.id.ad_title)
                .contentId(R.id.ad_content)
                .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.AD)
                .build();

        // small square size
        NendAdNativeClient mClient = new NendAdNativeClient(getApplicationContext(), SPOT_ID, API_KEY);

        mClient.loadAd(new NendAdNativeClient.Callback() {
            @Override
            public void onSuccess(final NendAdNative nendAdNative) {
                Log.i(TAG, "広告取得成功");
                nendAdNative.intoView(mAdContainer, mBinder);
                setPortraitPR();
                setBlank();
            }

            @Override
            public void onFailure(NendAdNativeClient.NendError nendError) {
                Log.i(TAG, "広告取得失敗 " + nendError.getMessage());
            }
        });
        setEndless(true);
    }

    /**
     * Get start coordinate
     */
    private int getMarqueeStartX() {
        return 0;
    }

    /**
     * Get end coordinate
     */
    private int getEndX() {
        return mWidthWithBlank;
    }

    /**
     * Set blank to title text
     */
    private void setBlank() {

        // get future title text size
        int nendTextWidth = (int) Layout.getDesiredWidth(mTextView.getText(), mTextView.getPaint());
        mWidthWithBlank = nendTextWidth + (mScrollViewWidth * 2);
        // first, set text view size
        mTextView.setWidth(mWidthWithBlank);

        // second, set padding
        mTextView.setPadding(mScrollViewWidth, 0, mScrollViewWidth, 0);
        startMarquee();

    }

    /**
     * Set portrait to PR text
     */
    private void setPortraitPR() {
        TextView pr = (TextView) findViewById(R.id.ad_pr);
        assert pr != null;
        float scale = getResources().getDisplayMetrics().density;
        int paddingDp = (int) (10 * scale + 0.5f);
        // set future pr size plus 10dp
        int prWidth = (int) Layout.getDesiredWidth(pr.getText(), pr.getPaint()) + paddingDp;
        pr.setWidth(prWidth / 2);
    }

    private void setEndless(boolean isEndless) {
        mIsEndlessLoop = isEndless;
    }

    /**
     * clear marquee
     */
    public void clearMarquee() {
        mCurrentX = getMarqueeStartX();
        mRepeatCount = 0;
        mThread = null;
    }

    /**
     * start marquee
     */
    public void startMarquee() {
        clearMarquee();
        mThread = new Thread(runnable);
        mThread.setName("marquee");
        mThread.start();
    }


    @Override
    protected void onPause() {
        super.onPause();
        // true to leaving the loop
        mThreadStopper = true;
        if (null != mThread && mThread.getName().equals("marquee") && !mThread.isInterrupted()) {
            mThread.interrupt();
        }
        clearMarquee();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mThreadStopper) {
            mThreadStopper = false;
            startMarquee();
        }
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            mScrollViewWidth = mMarqueeView.getWidth();
        } else {
            mThreadStopper = true;
            if (null != mThread && !mThread.isInterrupted()) {
                mThread.interrupt();
            }
            clearMarquee();
        }
    }

}

File: Sample/kotlin/src/main/java/net/nend/sample/kotlin/nativead/NativeMarqueeActivity.kt
package net.nend.sample.kotlin.nativead

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.Layout
import android.util.Log
import android.view.View
import android.widget.TextView
import kotlinx.android.synthetic.main.native_marquee.*
import net.nend.android.NendAdNative
import net.nend.android.NendAdNativeClient
import net.nend.android.NendAdNativeViewBinder
import net.nend.sample.kotlin.R
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_API_KEY_SMALL_SQUARE
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_LOG_TAG
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_SPOT_ID_SMALL_SQUARE

/**
 * Sample for like telop
 */
class NativeMarqueeActivity : AppCompatActivity() {

    private lateinit var marqueeView: MarqueeView
    private lateinit var textView: TextView
    private lateinit var adContainer: View

    private var threadStopper = false // for leaving the loop
    private var scrollViewWidth: Int = 0           // MarqueeView width

    /**
     * Get end coordinate
     */
    private var endX: Int = 0            // NendNativeTextView width
    private lateinit var thread: Thread          // runnable thread
    private var currentX: Int = 0                  // current text position
    private var isEndlessLoop: Boolean = false         // endless flag

    /**
     * Runnable variable for marquee
     */
    private val runnable = Runnable {
        // end point
        val endX = endX
        val textMoveSpeed = 10

        if (!threadStopper && !Thread.interrupted()) {

            currentX = marqueeStartX

            var beforeTime = System.currentTimeMillis()
            var afterTime: Long
            val fps = 30
            val frameTime = (1000 / fps).toLong()

            while (true) {
                if (!threadStopper && !Thread.interrupted()) {

                    if (currentX >= endX) {
                        currentX = marqueeStartX
                        if (!isEndlessLoop) {
                            break
                        }
                    }

                    currentX += textMoveSpeed

                    // UI control must on UI thread
                    runOnUiThread { marqueeView.smoothScrollTo(currentX, 0) }

                    // speed control
                    afterTime = System.currentTimeMillis()
                    val pastTime = afterTime - beforeTime
                    val sleepTime = frameTime - pastTime
                    if (sleepTime > 0) {
                        try {
                            Thread.sleep(sleepTime)
                        } catch (e: InterruptedException) {
                            Log.d(NATIVE_LOG_TAG, "thread interruption")
                            break
                        }

                    }
                    beforeTime = System.currentTimeMillis()
                }
            }
        }
    }

    /**
     * Get start coordinate
     */
    private val marqueeStartX: Int
        get() = 0

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.native_marquee)

        marqueeView = findViewById(R.id.horizontalScrollView) as MarqueeView
        textView = findViewById(R.id.ad_content) as TextView
        adContainer = findViewById(R.id.ad)

        val binder = NendAdNativeViewBinder.Builder()
                .adImageId(R.id.ad_image)
                .titleId(R.id.ad_title)
                .contentId(R.id.ad_content)
                .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.AD)
                .build()

        // small square size
        NendAdNativeClient(
                applicationContext, NATIVE_SPOT_ID_SMALL_SQUARE, NATIVE_API_KEY_SMALL_SQUARE).apply {
            loadAd(object : NendAdNativeClient.Callback {
                override fun onSuccess(nendAdNative: NendAdNative) {
                    Log.i(NATIVE_LOG_TAG, "広告取得成功")
                    nendAdNative.intoView(adContainer, binder)
                    setPortraitPR()
                    setBlank()
                }

                override fun onFailure(nendError: NendAdNativeClient.NendError) {
                    Log.i(NATIVE_LOG_TAG, "広告取得失敗: ${nendError.message}")
                }
            })
        }

        setEndless(true)
    }

    /**
     * Set blank to title text
     */
    private fun setBlank() {
        // get future title text size
        val nendTextWidth = Layout.getDesiredWidth(textView.text, textView.paint).toInt()
        endX = nendTextWidth + scrollViewWidth * 2
        // first, set text view size
        textView.width = endX
        // second, set padding
        textView.setPadding(scrollViewWidth, 0, scrollViewWidth, 0)
        startMarquee()
    }

    /**
     * Set portrait to PR text
     */
    private fun setPortraitPR() {
        val scale = resources.displayMetrics.density
        val paddingDp = (10 * scale + 0.5f).toInt()
        // set future pr size plus 10dp
        val prWidth = Layout.getDesiredWidth(ad_pr.text, ad_pr.paint).toInt() + paddingDp
        ad_pr.width = prWidth / 2
    }

    private fun setEndless(isEndless: Boolean) {
        isEndlessLoop = isEndless
    }

    /**
     * clear marquee
     */
    fun clearMarquee() {
        currentX = marqueeStartX
        threadStopper = true
    }

    /**
     * start marquee
     */
    fun startMarquee() {
        clearMarquee()
        threadStopper = false
        thread = Thread(runnable).apply {
            name = "marquee"
            start()
        }
    }

    override fun onPause() {
        super.onPause()
        // true to leaving the loop
        threadStopper = true
        if (thread.name == "marquee" && !thread.isInterrupted) {
            thread.interrupt()
        }
        clearMarquee()
    }

    override fun onResume() {
        super.onResume()
        if (threadStopper) {
            threadStopper = false
            startMarquee()
        }
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            scrollViewWidth = marqueeView.width
        } else {
            threadStopper = true
            if (!thread.isInterrupted) {
                thread.interrupt()
            }
            clearMarquee()
        }
    }
}